add some RealNest benchmarks

ZuleykhaPavlichenkova-TomTom opened the pull request:
This PR adds some benchmark queries to the RealNest JSON data as it discussed in the [issue](https://github.com/duckdblabs/duckdb-internal/issues/2421). 

For now there are queries to only one of the files - `hep-adl-ethz-Run2012B_SingleMu`, which is not a part of the PR, but can be found [here](https://github.com/cwida/RealNest/blob/main/sample-data/1024rows/hep-adl-ethz-Run2012B_SingleMu/data.jsonl) and should be placed in `duckdb_benchmark_data/` directory. 
Probably, the JSON file path should be changed after it becomes clear where to store RealNest test data (maybe somewhere outside of the `mount-point` of aws machine? )

ZuleykhaPavlichenkova-TomTom committed:
add some RealNest benchmarks

ZuleykhaPavlichenkova-TomTom committed:
ff

ff

ZuleykhaPavlichenkova-TomTom committed:
two more ff

taniabogatsch commented on benchmark/micro/json/RealNest/1_unnest_and_sum_one_json_struct.benchmark:
```diff
@@ -0,0 +1,17 @@
+# name: benchmark/micro/json/RealNest/1_unnest_and_sum_one_json_struct.benchmark
+# description: Create a view reading a JSON file, unnesting one of structures and summing its fields values
```
In addition to aggregating (`SUM`) the fields after unnesting, we can add a run computing the aggregate directly on the `STRUCT`.
```sql
SELECT i, SUM(s.a), SUM(s.b) FROM my_tbl GROUP BY i;
```

And a minor nit: I was confused because we do not create an SQL `VIEW` here; it is just a table.

hmeriann replied:
Thank you! I was too focused on unnesting everything))) With the applying aggregation directly on the `struct` it looks much cleaner and runs faster, but do we need a benchmark of this kind?

taniabogatsch replied:
Yes, we should keep some benchmarks for `UNNEST`. But others should not need any unnesting. :)

hmeriann replied:
Should we keep both variants: aggregating after unnesting and direct aggregating?

taniabogatsch replied:
Yes, I'd say so.

ZuleykhaPavlichenkova-TomTom committed:
rename a file

taniabogatsch commented on benchmark/micro/json/RealNest/2_create_table_combined_all_json_fields.benchmark:
I'm unsure if we need this benchmark: the join in `run` does not join any nested data. After the `load` step, all data is flat, so this benchmarks our usual join performance.

hmeriann replied:
In this one I wasn't sure about too. But maybe it would be more useful to create table unnesting from read_json directly, what do you think?
like just run:
```run
CREATE OR REPLACE TABLE pv AS SELECT UNNEST(PV) from (SELECT * FROM read_json('duckdb_benchmark_data/hep-adl-ethz-Run2012B_SingleMu/data.jsonl'));
```

taniabogatsch replied:
I think @Tmonster's comment below makes sense. That way, we can turn this into a big `UNNEST` benchmark.

hmeriann replied:
Nice!

taniabogatsch commented on benchmark/micro/json/RealNest/3_select_avg_from_unnested_structure.benchmark:
```diff
@@ -0,0 +1,21 @@
+run
+SELECT puId, avg(pt), avg(eta), avg(phi), avg(mass), avg(btag)
+  FROM (
+      SELECT UNNEST(Jet, recursive:=true) AS unnested_jet
```
Is `Jet` a list? Are `mass` and `eta` nested types?

hmeriann replied:
`Jet` is a `list` type, yeah.
 `mass` and `eta` are nested as well as `pt`, `btag`, `phi`

taniabogatsch replied:
Great, then this also benchmarks aggregating nested values 👍

taniabogatsch commented on benchmark/micro/json/RealNest/4_create_table_combined_all_json_fields.benchmark:
```diff
@@ -0,0 +1,22 @@
+    count(*) FILTER (charge > 0) as neutral,
+    count(*) FILTER (pt - relIso_all < decayMode) as odds,
+    count(*) FILTER (idIsoVLoose != idIsoLoose) as idIsoMatch
+FROM (SELECT UNNEST(Tau, recursive:=true) AS tau FROM Run2012B_SingleMu)
```
Is `Tau` a list? Are the fields in the `SELECT` clause nested?

taniabogatsch commented on benchmark/micro/json/RealNest/5_list_filter_on_unnested_structure.benchmark:
```diff
@@ -0,0 +1,21 @@
+
+run
+SELECT list_filter(
+        [pt, eta, phi, mass, pfRelIso03_all, pfRelIso04_all, dxy, dxyErr, jetIdx, genPartIdx],
```
Are there lists in the data set? It would be nice to benchmark `list_filter` without the overhead of first creating the list like here. Nonetheless, we should keep this benchmark 👍

hmeriann replied:
There are multiple lists in this data set: Muon, Electron, Tau, Photon, Jet.
And yes, it is interesting to try to benchmark them without creating the list, I'll see if I could do that)

taniabogatsch replied:
Great. :)

taniabogatsch commented on benchmark/micro/json/RealNest/4_filter_unnested_struct.benchmark:
```diff
@@ -0,0 +1,22 @@
+    count(*) FILTER (charge > 0) as neutral,
+    count(*) FILTER (pt - relIso_all < decayMode) as odds,
+    count(*) FILTER (idIsoVLoose != idIsoLoose) as idIsoMatch
+FROM (SELECT UNNEST(Tau, recursive:=true) AS tau FROM Run2012B_SingleMu)
```
Is `Tau` a list? Are any of the columns in the `SELECT` clause nested types?

hmeriann replied:
`Tau` is a `list` and yes, all the columns are nested here

taniabogatsch replied:
Nice 👍 
So this benchmark contains filters on nested values, and `benchmark/micro/json/RealNest/5_list_filter_on_unnested_structure.benchmark` aggregates them. That's cool.

taniabogatsch commented on benchmark/micro/json/RealNest/6_select_sum_from_unnested_structure.benchmark:
I'm unsure if we need to keep this benchmark, as it also benchmarks `UNNEST`, which we already do in the others.

hmeriann replied:
I can replace it with, for example, ```select list_sort(flatten(list_slice(list(Jet), 150, 200, 3)), 'DESC') from Run2012B_SingleMu;
```, if it is better one.

taniabogatsch replied:
Yes, perfect!

taniabogatsch reviewed (changes requested):
Hey @hmeriann! Thanks for starting to work on this! :) I've left some comments and added a more extensive comment to the original issue.

Currently, this PR focuses heavily on the `UNNEST`. In future PRs, it would be nice to see more operations directly on the nested types without unnesting them.

Tmonster reviewed (commented):
just the one comment for now, will continue tomorrow

ZuleykhaPavlichenkova-TomTom committed:
Change according to the comments

ZuleykhaPavlichenkova-TomTom committed:
ff

ZuleykhaPavlichenkova-TomTom committed:
Add multilayer unnest from loaded table and directly

ZuleykhaPavlichenkova-TomTom committed:
Count len of list concatenating two list ranges

ZuleykhaPavlichenkova-TomTom committed:
Add a query which creates a list of unnested texts, counts texts string's length and returns a number of unique values

ZuleykhaPavlichenkova-TomTom committed:
ff

ZuleykhaPavlichenkova-TomTom committed:
rename files, change descriptions and some queries

ZuleykhaPavlichenkova-TomTom committed:
count map keys benchmark

ZuleykhaPavlichenkova-TomTom committed:
add list_intersect benchmark

ZuleykhaPavlichenkova-TomTom committed:
ff

ZuleykhaPavlichenkova-TomTom committed:
add load.sql file

ZuleykhaPavlichenkova-TomTom committed:
move realnest benchmarks out of the micro directory

ZuleykhaPavlichenkova-TomTom committed:
ff

ZuleykhaPavlichenkova-TomTom committed:
load.sql

ZuleykhaPavlichenkova-TomTom committed:
fix mixed up filenames

ZuleykhaPavlichenkova-TomTom committed:
Add queries with array_agg, list_reduce, list_transform, list_sort

Tmonster commented on benchmark/realnest/3_list_transform_plus_list_aggregate.benchmark:
```diff
@@ -0,0 +1,24 @@
+  avg(list_aggregate(list_transform(Jet, x -> x.btag), 'avg')) as avg_btag
+  FROM Run2012B_SingleMu
+  GROUP BY npvs
+  HAVING avg_mass > avg_eta
```
I don't think we need a having clause here

Tmonster commented on benchmark/realnest/8_list_intersect_hashtags.benchmark:
```diff
@@ -0,0 +1,20 @@
+run
+WITH A AS (
+    SELECT rowid, data.text AS text, list_transform(data.entities.hashtags, x -> x.tag) AS hashtag_list from twitter)
+    SELECT left_.text, right_.text, len(list_intersect(left_.hashtag_list, right_.hashtag_list)) AS common_hashtags
```
```suggestion
SELECT left_.text, right_.text, len(list_intersect(left_.hashtag_list, right_.hashtag_list)) AS common_hashtags
```

Tmonster commented on benchmark/realnest/9_select_sum_from_unnested_structure.benchmark:
```diff
@@ -0,0 +1,20 @@
+# name: benchmark/realnest/9_aggregate_nested_structs.benchmark
```
I think the name here should be, `09_array_agg.benchmark`

Tmonster commented on benchmark/realnest/3_list_transform_plus_list_aggregate.benchmark:
```diff
@@ -0,0 +1,24 @@
+
+run
+SELECT PV.npvs as npvs, 
+  avg(list_aggregate(list_transform(Jet, x -> x.pt), 'avg')) as avg_pt,
```
If I'm not mistaken, the dataset has many more columns with a list of structs.
To really stress test list_transform + list_aggregate you could also caluclate some aggregates for the following columns

`Muon`, `Electron`, `Tau`, and `Photon`

hmeriann replied:
Added all of them to the `04_list_transform_plus_list_aggregate` benchmark

Tmonster reviewed (commented):
Thanks! 
This looks good for now. Let's wait until Tania gets back to see if we should add more. 

Also, could you rename all of the single digit benchmarks so that they are prefixed with a `0`?

I.e, instead of `1_list_sort_text.benchmark`  it's `01_list_sort_text.benchmark`, then they show up as sorted in the file system

ZuleykhaPavlichenkova-TomTom committed:
rename files

ZuleykhaPavlichenkova-TomTom committed:
add another map benchmark and clean up

taniabogatsch commented on benchmark/realnest/09_array_agg.benchmark:
```diff
@@ -0,0 +1,26 @@
+
+load benchmark/realnest/load.sql
+
+run
```
Is it intentional to have two `SELECT` in the same benchmark here?

hmeriann replied:
Oh, sorry, it was accidentally. I'll make sure to remove the extra `SELECT`

taniabogatsch reviewed (commented):
Looks good from my side now 😄 - I only added a small comment.

ZuleykhaPavlichenkova-TomTom committed:
remove an extra statement

ZuleykhaPavlichenkova-TomTom committed:
add list_transform plus list_aggregate to all the columns from the Muon, Electron, Tau and Photon columns with doubles and bigints

ZuleykhaPavlichenkova-TomTom committed:
fix indentation

ZuleykhaPavlichenkova-TomTom committed:
add multi join nested data with filtering benchmark

hmeriann commented on benchmark/micro/json/RealNest/2_create_table_combined_all_json_fields.benchmark:
```diff
@@ -0,0 +1,38 @@
+CREATE VIEW unnested_jet AS (SELECT rowid, UNNEST(Jet, recursive:=true) AS jet FROM Run2012B_SingleMu);
+
+run
+CREATE OR REPLACE TABLE combined AS (SELECT unnested_hlt.*,
```
Added a `13_multi_join_nested_data_with_filtering` benchmark

Tmonster commented on benchmark/realnest/01_aggregate-on-one-deep-struct-members.benchmark:
```diff
@@ -0,0 +1,19 @@
+
+require httpfs
+
+cache real_nest.duckdb
```
this `cache real_nest.duckdb` needs to be in every `.benchmark` file. Otherwise the data gets downloaded again for every benchmark

Tmonster reviewed (commented):
Also can you add this as a regression test to the `Regression.yml`
You can add it after the `regression test csv` step that looks like this
```
    - name: Regression Test CSV
      if: always()
      shell: bash
      run: |
        python scripts/regression_test_runner.py --old=duckdb/build/release/benchmark/benchmark_runner --new=build/release/benchmark/benchmark_runner --benchmarks=.github/regression/csv.csv --verbose --threads=2
```

Tmonster commented on benchmark/realnest/05_list_filters.benchmark:
```diff
@@ -0,0 +1,28 @@
+    sum(len(list_filter(Photon, x -> x.mass > 0))) as photon,
+    sum(len(list_filter(Jet, x -> x.puId != false))) as jet,
+
+FROM Run2012B_SingleMu
```
this also needs to be removed

Tmonster commented on benchmark/realnest/08_count_map_keys.benchmark:
```diff
@@ -0,0 +1,19 @@
+
+run
+SELECT keys, count(*) mentions FROM
+    (SELECT UNNEST(map_keys(tags)) AS keys FROM daylight)
```
also the name of the table here needs to match the table name in `load.sql`. So either `daylight` here needs to change to `open_street_map` or `open_street_map` in the load.sql needs to change to `daylight`

ZuleykhaPavlichenkova-TomTom committed:
add cache real_nest.duckdb

ZuleykhaPavlichenkova-TomTom committed:
fix table name and add a 14_list_sort_flatten_slice.benchmark from the comment

ZuleykhaPavlichenkova-TomTom committed:
add realnest to the Regression.yml

hmeriann commented:
> Also can you add this as a regression test to the `Regression.yml`

Done

ZuleykhaPavlichenkova-TomTom committed:
fix a typo

Tmonster commented on benchmark/realnest/12_map_two_list_values.benchmark:
```diff
@@ -0,0 +1,19 @@
+load benchmark/realnest/load.sql
+
+run
+SELECT map([gh_issue.issue.user.login],[gh_pull.pull_request.body]) 
```
looking at this one again, I'm not sure what we are benchmarking. In this case 
`select [gh_issue.issue.user.login] as lst_login, [gh_pull.pull_request.body] as lst_body FROM gh_pull, gh_issue WHERE gh_issue.issue.user = gh_pull.pull_request.head.repo.owner
  GROUP BY ALL limit 5;` returns the following
```
┌────────────────────┬─────────────────────────────────────────────────────────┐
│     lst_login      │                        lst_body                         │
│     varchar[]      │                        varchar[]                        │
├────────────────────┼─────────────────────────────────────────────────────────┤
│ [matsumo0922]      │ [# Related links\r\n- N/A\r\n\r\n# How?\r\n- バージョ…  │
│ [TiwariRahul16]    │ [## PR Description 📜\r\n\r\nAdded [New game] Rock_pa…  │
│ [LashaJini]        │ [NULL]                                                  │
│ [Thraximundar2772] │ [NULL]                                                  │
│ [SimVac]           │ [close #42 ]                                            │
└────────────────────┴─────────────────────────────────────────────────────────┘
```

This basically means the benchmarked query is calling map on two lists of length 1 over and over again. In addition, we are benchmarking joining on useres again, which is done pretty heavily in `benchmark/realnest/13_multi_join_nested_data_with_filtering.benchmark`

I think what would be nice is to find a dataset with on table that has two list columns have of the same length (or we can make them the same length with `list_resize`) and then we call `map(list_1, list_2)`. Then map is called on two lists with. multiple different list lengths. 

@taniabogatsch what do you think?

taniabogatsch replied:
Yes, that makes a lot of sense. This is also why we resized the lists in one of the views?

Tmonster commented on benchmark/realnest/14_list_sort_flatten_slice.benchmark:
```diff
@@ -0,0 +1,21 @@
+load benchmark/realnest/load.sql
+
+run
+SELECT list_sort(flatten(list_slice(list(Jet), 150, 200, 3)), 'ASC'),
```
Same here, I don't quite understand what is getting benchmarked. 

`list(Jet), list(Muon), list(Photon), list(Tau)` all create on big list from the row lists. I think it's better if we make list_slice operate on a row level basis. You can create ta copy of the run2012B_singleMu table by placing the following query in the load.sql

```
create or replace table single_mu_lists as select * REPLACE(list_resize(Jet, 10, NULL) as Jet, list_resize(Muon, 10, NULL) as Muon, list_resize(Photon, 10, NULL) as Photon, list_resize(Tau, 10, NULL) as Tau) from run2012B_singleMu;
```

The benchmark query could then look like this
```
SELECT list_sort(list_slice(Jet, 2, 9), 'ASC'),
   list_sort(list_slice(Muon, 1, 6), 'DESC'),
    list_sort(list_slice(Photon, 1, 3), 'ASC'),
    list_sort(list_slice(Tau, 5, 9), 'DESC')
FROM run2012B_singleMu;
```

and you'd get the same number of rows as in the input table. @taniabogatsch would this make more sense? Or should the benchmark just stay as is?

taniabogatsch replied:
I think a dedicated benchmark for slicing is excellent. Also, we can omit the `list_sort`, since we are already benchmarking it in a different file (IIRC).

taniabogatsch replied:
Resolved

ZuleykhaPavlichenkova-TomTom committed:
fix map benchmark, add precondition query to the load.sql file

ZuleykhaPavlichenkova-TomTom committed:
separate benchmarks for list_slice and list_sort

ZuleykhaPavlichenkova-TomTom committed:
ff

taniabogatsch reviewed (commented):
Hi @hmeriann, I had a look at the changes and added a few more comments. Great work!

ZuleykhaPavlichenkova-TomTom committed:
replace VIEWS with TABLES

ZuleykhaPavlichenkova-TomTom committed:
add more list_sorts with a different null ordering

taniabogatsch commented:
A small note on this: https://github.com/duckdb/duckdb/pull/13345/commits/d8a1b8b3cc8b65f831c04bbb2c20739269711d43

We should keep/add 1-2 dedicated `UNNEST` benchmarks.

ZuleykhaPavlichenkova-TomTom committed:
add a step (third parameter), and a reverse (negative parameter)

hmeriann commented:
> We should keep/add 1-2 dedicated `UNNEST` benchmarks.

I'll add the benchmarks with queries performing unnest and keep the `load.sql` as it is now, right?

Tmonster commented on benchmark/realnest/load.sql:
Responding here so context isn't lost. The `views` were only created for one benchmark (`benchmark/realnest/03_create_table_combined_all_json_fields.benchmark`)meant to test unnest. The creation of the views were included in the `load` script so that the benchmark wouldn't be cluttered with multiple subqueries with unnest

taniabogatsch replied:
Oh! Then I misread that entirely - my bad! In that case, the views make sense.

Tmonster commented on benchmark/realnest/03_create_table_combined_all_json_fields.benchmark:
```diff
@@ -0,0 +1,33 @@
+    unnested_jet.*
+    FROM unnested_hlt
+    LEFT JOIN unnested_pv ON unnested_hlt.rowid = unnested_pv.rowid
+    LEFT JOIN unnested_met ON unnested_hlt.rowid = unnested_met.rowid
```
So for this benchmark, the `unnested_*` tables have already been unnested in the load step. 
I think it might be better if they are unnested in the query to avoid confusion. 
So maybe the query could look like
```
Create or replace table combined as 
(select 
    unnested_pv.*
    unnested_met.*
    ....
    FROM 
        (SELECT rowid, UNNEST(htl) as hlt FROM run2012B_singleMu) as unnested_hlt 
        LEFT JOIN
        (SELECT rowid, UNNEST(pv) AS pv FROM run2012B_singleMu) as unnested_pv ON  unnested_hlt.row_id = unnested_pv.row_id 
        LEFT JOIN 
        (SELECT rowid, UNNEST(met) AS met FROM run2012B_singleMu) as unnested_met ON  unnested_hlt.row_id = unnested_met.row_id 
        ....
```

That way it is clear to future readers where the unnesting is happening

Tmonster replied:
You can then remove all of the `create table` statements in the `load.sql` that are basically doing the same thing

ZuleykhaPavlichenkova-TomTom committed:
change case 03 create table from unnested structs

ZuleykhaPavlichenkova-TomTom committed:
list aggregate benchmarks

ZuleykhaPavlichenkova-TomTom committed:
ff

ZuleykhaPavlichenkova-TomTom committed:
fix map benchmark and two benchmarks that query the same table

ZuleykhaPavlichenkova-TomTom committed:
make load run2012B_singleMu from s3

ZuleykhaPavlichenkova-TomTom committed:
set limit ro bechnmark 03, fix the load.sql

ZuleykhaPavlichenkova-TomTom committed:
less data for the singleMu

ZuleykhaPavlichenkova-TomTom committed:
more data for the singleMu, but with lower limit

ZuleykhaPavlichenkova-TomTom committed:
run only the RealNest benchmarks

ZuleykhaPavlichenkova-TomTom committed:
run only the RealNest benchmarks

ZuleykhaPavlichenkova-TomTom committed:
don't set up old benchmarks

ZuleykhaPavlichenkova-TomTom committed:
larger data sets for open street map, github and twitter

ZuleykhaPavlichenkova-TomTom committed:
fix create table from osm

taniabogatsch commented on benchmark/realnest/load.sql:
```diff
@@ -0,0 +1,23 @@
+create table gh_issue as select * from read_json('https://github.com/cwida/RealNest/raw/cdaa85652bf187226706b321c29597673b2b7d86/sample-data/100mib/gharchive-IssuesEvent/data.jsonl');
+create table gh_pull as select * from read_json('https://github.com/cwida/RealNest/raw/cdaa85652bf187226706b321c29597673b2b7d86/sample-data/100mib/gharchive-PullRequestEvent/data.jsonl');
+
+CREATE VIEW single_mu_lists AS SELECT * REPLACE(list_resize(Jet, 10, NULL) as Jet, list_resize(Muon, 10, NULL) as Muon, list_resize(Photon, 10, NULL) as Photon, list_resize(Tau, 10, NULL) as Tau) 
```
Yes, we need the same list length for the `MAP` function.

taniabogatsch commented on .github/workflows/Regression.yml:
Let's not forget to enable all other benchmarks before merging. :)

hmeriann replied:
resolved

taniabogatsch commented on benchmark/realnest/01_aggregate-on-one-deep-struct-members.benchmark:
```diff
@@ -0,0 +1,19 @@
+# name: benchmark/realnest/01_aggregate-on-one-deep-struct-members.benchmark
```
nit: `aggregate-first-level-struct-members`

taniabogatsch commented on benchmark/realnest/02_list_sort_text.benchmark:
```diff
@@ -0,0 +1,17 @@
+load benchmark/realnest/load.sql
+
+run
+SELECT list_sort(list_transform(body_text, x -> x.text)) FROM cord;
```
`list_sort(body_text)`, if the runtime does not explode, which it should not.

hmeriann replied:
resolved

taniabogatsch commented on benchmark/realnest/02_list_sort_text.benchmark:
```diff
@@ -0,0 +1,17 @@
+# name: benchmark/realnest/02_list_sort_text.benchmark
```
`list_sort.benchmark`

hmeriann replied:
resolved

taniabogatsch commented on benchmark/realnest/03_create_table_from_unnested_structs.benchmark:
```diff
@@ -0,0 +1,43 @@
+# name: benchmark/realnest/03_create_table_from_unnested_structs.benchmark
+# description: Create a table combining all the fields from a JSON file, unnesting all structures
```
`Create a table by unnesting and joining the structs in a JSON file`

hmeriann replied:
resolved

taniabogatsch commented on benchmark/realnest/03_create_table_from_unnested_structs.benchmark:
```diff
@@ -0,0 +1,43 @@
+    (SELECT rowid, UNNEST(Photon, recursive:=true) AS photon FROM run2012B_singleMu) as unnested_photon ON unnested_hlt.rowid = unnested_photon.rowid
+    LEFT JOIN
+    (SELECT rowid, UNNEST(Jet, recursive:=true) AS jet FROM run2012B_singleMu) as unnested_jet ON unnested_hlt.rowid = unnested_jet.rowid
+LIMIT 100000
```
Maybe we can open an issue in the internal repository (with a reproduction) to understand if the benchmark `timeout` for more rows is expected.

taniabogatsch commented on benchmark/realnest/04_list_transform_plus_list_aggregate.benchmark:
```diff
@@ -0,0 +1,62 @@
+# name: benchmark/realnest/04_list_transform_plus_list_aggregate.benchmark
```
`list_transform_and_list_aggregate`

taniabogatsch commented on benchmark/realnest/04_list_transform_plus_list_aggregate.benchmark:
```diff
@@ -0,0 +1,62 @@
+# name: benchmark/realnest/04_list_transform_plus_list_aggregate.benchmark
+# description: select average from transformed list and group by, having
```
Calculate different list averages after transforming the list. The `AVG` serves to limit the result set size and has no significant impact on performance.

We should be able to remove the `GROUP BY`.
```sql
D select avg(len(j)) from l;
┌─────────────┐
│ avg(len(j)) │
│   double    │
├─────────────┤
│         2.0 │
└─────────────┘
D from l;
┌──────────┐
│    j     │
│ int32[]  │
├──────────┤
│ [42, 43] │
└──────────┘
```

taniabogatsch commented on benchmark/realnest/05_list_filters.benchmark:
```diff
@@ -0,0 +1,28 @@
+# name: benchmark/realnest/05_list_filters.benchmark
```
`list_filter.benchmark`

taniabogatsch commented on benchmark/realnest/07_select_text.benchmark:
```diff
@@ -0,0 +1,18 @@
+# name: benchmark/realnest/07_select_text.benchmark
```
`07_list_unique_on_transformed_and_aggregated_list.benchmark`

taniabogatsch commented on benchmark/realnest/07_select_text.benchmark:
```diff
@@ -0,0 +1,18 @@
+# name: benchmark/realnest/07_select_text.benchmark
+# description: Creates a list of unnested texts, counts texts string's length and returns a number of unique values
```
Creates a list by unnesting and re-aggregating it, then performs `list_transform` and `list_unique` on that list.

taniabogatsch commented on benchmark/realnest/08_count_map_keys.benchmark:
```diff
@@ -0,0 +1,20 @@
+SELECT keys, count(*) mentions FROM (
+    SELECT UNNEST(map_keys(tags)) AS keys 
+FROM open_street_map) GROUP BY keys 
+ORDER BY mentions DESC LIMIT 20;
```
Maybe we can increase the limit slightly.

taniabogatsch commented on benchmark/realnest/10_list_intersect_hashtags.benchmark:
We should remove this and turn it into an issue. Later, we can add it back in.

hmeriann replied:
https://github.com/duckdblabs/duckdb-internal/issues/2991

taniabogatsch commented on benchmark/realnest/12_map_list_values.benchmark:
```diff
@@ -0,0 +1,23 @@
+    map(ph_pt, ph_eta)
+FROM singleMu
+ORDER BY ALL DESC
+LIMIT 10;
```
Double-check if we can increase the limit.

taniabogatsch commented on benchmark/realnest/16_most_common_list_aggregates.benchmark:
```diff
@@ -0,0 +1,82 @@
+
+run
+SELECT 
+  list_aggregate(tau_pt, 'list'),
```
Also in the other benchmarks, you can rewrite some of the `list_aggregates` to their respective rewrites.
`list`, `list_sum`, ... What you prefer. 😄

taniabogatsch reviewed (changes requested):
- Once all the queries are ready (which they mostly are) you can do a final pass to unify the formatting and names.

ZuleykhaPavlichenkova-TomTom committed:
add cord 10k dataset

ZuleykhaPavlichenkova-TomTom committed:
add cord 10k dataset

ZuleykhaPavlichenkova-TomTom committed:
enable other benchmarks in Regression.yml

ZuleykhaPavlichenkova-TomTom committed:
rename q01

ZuleykhaPavlichenkova-TomTom committed:
rename and change q02

ZuleykhaPavlichenkova-TomTom committed:
rename and clean up q04

ZuleykhaPavlichenkova-TomTom committed:
update description

ZuleykhaPavlichenkova-TomTom committed:
rename q05

ZuleykhaPavlichenkova-TomTom committed:
rename q07

ZuleykhaPavlichenkova-TomTom committed:
remove limit

ZuleykhaPavlichenkova-TomTom committed:
remove benchmark/realnest/10_list_intersect_hashtags.benchmark

ZuleykhaPavlichenkova-TomTom committed:
remove limit from benchmark/realnest/12_map_list_values.benchmark

ZuleykhaPavlichenkova-TomTom committed:
ff

ZuleykhaPavlichenkova-TomTom committed:
fix the realnest.csv

hmeriann commented:
Here you can see the timings from the [CI run](https://github.com/hmeriann/test_/actions/runs/10768856809/job/29858810660) on the GH machine.

Tmonster commented on benchmark/realnest/01_aggregate-first-level-struct-members.benchmark:
```diff
@@ -0,0 +1,20 @@
+load benchmark/realnest/load.sql
+
+run
+SELECT PV.npvs, sum(PV.x) AS sum_x, sum(PV.y) AS sum_y, sum(PV.z) AS sum_z, avg(MET.pt) as avg_pt, min(MET.phi) as min_phi, max(MET.sumet) as max_sumet
```
nit: every selected expression can be on a separate line

Tmonster reviewed (commented):
Looks good except the the `Regression.yml`

ZuleykhaPavlichenkova-TomTom committed:
some fixes of the alignment

ZuleykhaPavlichenkova-TomTom committed:
fix alignment for the Regression.yml

ZuleykhaPavlichenkova-TomTom committed:
set BUILD_JSON: 1 flag

Tishj committed:
these methods are static because they create a BufferHandle from the 'shared_ptr<BlockHandle>', but we can just use 'shared_from_this()' instead

Tishj committed:
use optional_ptr and reference in BufferManager::Allocate so the code is less confusing

Tishj committed:
remove the optional shared_ptr<BlockHandle> from Allocate(...), it is instead retrieved afterwards from the BufferHandle

ZuleykhaPavlichenkova-TomTom committed:
fix indentation

Tishj committed:
add patch for Spatial

Tishj committed:
fix up the test and verify that the entries are set correctly by looping through it entry-wise

hmeriann commented:
> Looks good except the the `Regression.yml`

Yeah, that was because of the indentation. Fixed it

pdet committed:
fix comment prompt

pdet committed:
only print quote on prompt once

ZuleykhaPavlichenkova-TomTom committed:
fix indentation

ZuleykhaPavlichenkova-TomTom committed:
rearrange the order of the becnhmarks to see if the same benchmark fails

ZuleykhaPavlichenkova-TomTom committed:
experiment with the timeout before the benchmarks run

ZuleykhaPavlichenkova-TomTom committed:
increase the timeout, set it to 5 sec

ZuleykhaPavlichenkova-TomTom committed:
increase the timeout, set it to 5 sec

ZuleykhaPavlichenkova-TomTom committed:
correct path to the data set

ZuleykhaPavlichenkova-TomTom committed:
wget realnest data before benchmarks run

ZuleykhaPavlichenkova-TomTom committed:
change back the order of the benchmarks in the csv file

ZuleykhaPavlichenkova-TomTom committed:
discard changes in the load script

ZuleykhaPavlichenkova-TomTom committed:
Merge branch 'main' into real-nest-benchmarks

Tmonster commented on scripts/regression_test_runner.py:
```diff
@@ -95,6 +95,8 @@ def run_benchmark(runner, benchmark):
     else:
         timeout_seconds = 600
     try:
+        import time
```
these can be removed as well

ZuleykhaPavlichenkova-TomTom committed:
mkdir

ZuleykhaPavlichenkova-TomTom committed:
ff

ZuleykhaPavlichenkova-TomTom committed:
remove timout

ZuleykhaPavlichenkova-TomTom committed:
fix paths for duckdb_benchmark_data

ZuleykhaPavlichenkova-TomTom committed:
mkdir -p

Tmonster reviewed (approved):
Looks good to me! Thanks 👍

taniabogatsch commented on benchmark/realnest/load.sql:
```diff
@@ -0,0 +1,27 @@
+attach 'https://duckdb-blobs.s3.amazonaws.com/data/realnest/cord_10k.duckdb' as cord (READ_ONLY);
+create table cord as select * from cord.cord;
```
Uppercase SQL keywords like `CREATE TABLE` and `AS`.

taniabogatsch reviewed (approved):
Looks good to me now! Great work.

Tishj committed:
move the definition into a header file so it can be further derived

ZuleykhaPavlichenkova-TomTom committed:
Merge branch 'main' into real-nest-benchmarks

ZuleykhaPavlichenkova-TomTom committed:
uppercase SQL keywords in load.sql

Mytherin committed:
Fix #14077: correctly reset next pointer when reconstructing new row group segment tree after vacuum

Mytherin committed:
No print

Mytherin committed:
Write bytes_read

Tmonster committed:
added the answer

szarnyasg committed:
CSV sniffer: Add spaces to make the error message consistent

szarnyasg committed:
CSV sniffer: Add more separators to error messages

Mytherin committed:
[Dev] Move `EnumTypeInfoTemplated` definition into a `hpp` file (#14038)

To further override `EnumTypeInfoTemplated` it needs to be either
vendored (bad) or we move it to a header.
Relevant `pg_duckdb` PR: <https://github.com/duckdb/pg_duckdb/pull/193>

Mytherin committed:
Fix #14077: correctly reset next pointer when reconstructing new row group segment tree after vacuum (#14092)

Fixes #14077

Mytherin committed:
Format CSV error messages (#14097)

Add some spaces and other seperators.

I spotted in the docs that the formatting of the sniffer's error
messages is inconsistent, see
https://duckdb.org/docs/data/csv/reading_faulty_csv_files#anatomy-of-a-csv-error

```
  file= people.csv
  delimiter = , (Auto-Detected)
  quote = " (Auto-Detected)
  escape = " (Auto-Detected)
  new_line = \r\n (Auto-Detected)
  header = true (Auto-Detected)
  skip_rows = 0 (Auto-Detected)
  date_format = (DD-MM-YYYY) (Auto-Detected)
  timestamp_format =  (Auto-Detected)
  null_padding=0
  sample_size=20480
  ignore_errors=false
  all_varchar=0
```

This PR adds a bunch of whitespace and also other separators for similar
error messages.

Mytherin committed:
Fix the answer file for tpcds q67 at sf100 (#14096)

Seems like some `NULL` values snuck in there. This was causing issues on
the weekly regression runner

ZuleykhaPavlichenkova-TomTom committed:
Merge branch 'main' into real-nest-benchmarks

hannes commented:
thanks!
